opentasks-provider/src/main/java/org/dmfs/provider/tasks/utils/InstanceValuesIterable.java

```diff
@@ -53,28 +56,18 @@ public InstanceValuesIterable(TaskAdapter taskAdapter)
     public Iterator<Single<ContentValues>> iterator()
     {
         Optional<DateTime> start = new NullSafe<>(mTaskAdapter.valueOf(TaskAdapter.DTSTART));
-        Optional<DateTime> due = new NullSafe<>(mTaskAdapter.valueOf(TaskAdapter.DUE));
-
-        Optional<Duration> effectiveDuration = new FirstPresent<>(
+        // effective due is either the actual due, start + duration or absent
+        Optional<DateTime> effectiveDue = new FirstPresent<>(
                 new Seq<>(
-                        new NullSafe<>(mTaskAdapter.valueOf(TaskAdapter.DURATION)),
-                        new Zipped<>(start, due, new DateTimeDurationBiFunction())));
-
-        // TODO: implement support for recurrence, for now we only return the first instance
-        return new SingletonIterator<Single<ContentValues>>(
-                new InstanceDateTimeData(start, due, effectiveDuration, new NullSafe<>(mTaskAdapter.valueOf(TaskAdapter.ORIGINAL_INSTANCE_TIME))));
-    }
+                        new NullSafe<>(mTaskAdapter.valueOf(TaskAdapter.DUE)),
+                        new Zipped<>(start, new NullSafe<>(mTaskAdapter.valueOf(TaskAdapter.DURATION)), DateTime::addDuration)));
 
+        Single<ContentValues> baseData = new Enduring(new DueDated(effectiveDue, new StartDated(start, new VanillaInstanceData())));
```

Comment:
Interesting to see that `Single`s can be used this way as well, to have this kind of composable factory.  Looks nice.

Reply:
It's a bit of a stretch because `ContentValues` is mutable and using `Frozen` would break this pattern.